image/refs/interface.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package refs implements generic name-based reference access.
+package refs
+
+import (
+	"github.com/opencontainers/image-spec/image/structure"
+)
+
+// Engine represents a name-based reference storage engine.
+type Engine interface {
+
+	// Put adds a new reference to the store.
+	Put(name string, descriptor *structure.Descriptor) (err error)
+
+	// Get returns a reference from the store.  Returns io.EOF if the
+	// name is not found.
+	Get(name string) (descriptor *structure.Descriptor, err error)
+
+	// List returns a slice of available names from the store.
+	//
+	// Results are sorted alphabetically.
+	//
+	// Arguments:
+	//
+	// * prefix: limits the result set to names starting with that
+	//   value.
+	// * size: limits the length of the result set to the first 'size'
+	//   matches.  A value of -1 means "all results".
+	// * from: shifts the result set to start from the 'from'th match.
+	//
+	// For example, a store with names like:
+	//
+	// * 123
+	// * abcd
+	// * abce
+	// * abcf
+	// * abcg
+	//
+	// will have the following call/result pairs:
+	//
+	// * List("", -1, 0) -> ["123", "abcd", "abce", "abcf", "abcg"].
+	// * List("", 2, 0) -> ["123", "abcd"]
+	// * List("", 2, 1) -> ["abcd", "abce"]
+	// * List("abc", 2, 1) -> ["abce", "abcf"].
+	List(prefix string, size int, from int) (names []string, err error)
```

Comment:
There might be better API here (for large ref sets) if we return a channel, so folks can launch a `List` goroutine and iterate over names without having to construct temporary slices.  That opens the door to iterating over all names without paginating, because you wouldn't have to hold a huge `names` slice in memory at once.

I think we want to keep the pagination arguments though, so callers can efficiently jump to a particular position or launch several parallel `List` calls to walk the namespace (assuming the backend efficiently implements the pagination arguments).